File: funcscript/parser/syntax/FuncScriptParser.GetInt.cs
using FuncScript.Model;
namespace FuncScript.Core
{
    public partial class FuncScriptParser
    {
        record GetIntResult(string IntVal, ParseNode ParseNode, int NextIndex)
            :ParseResult(ParseNode, NextIndex);

        static GetIntResult GetInt(ParseContext context, bool allowNegative, int index)
        {
            ParseNode parseNode = null;
            int i = index;
            if (allowNegative)
                i = GetLiteralMatch(context, i, "-").NextIndex;

            var i2 = i;
            while (i2 < context.Expression.Length && char.IsDigit(context.Expression[i2]))
                i2++;

            if (i == i2)
            {
                return new GetIntResult(null, parseNode, index);
            }

            i = i2;

            string intVal = context.Expression.Substring(index, i - index);
            parseNode = new ParseNode(ParseNodeType.LiteralInteger, index, i - index);
            return new GetIntResult(intVal, parseNode, i);
        }
    }
}

File: ts/core/src/parser/syntax/FuncScriptParser.GetInt.ts
import { ParseContext, ParseNode, ParseNodeType, ParseResult, SyntaxErrorData } from "../FuncScriptParser.Main";
import { GetLiteralMatch } from "./FuncScriptParser.GetLiteralMatch";

export interface GetIntResult extends ParseResult {
    IntVal: string | null;
}

export function GetInt(context: ParseContext, allowNegative: boolean, index: number): GetIntResult {
    let parseNode: ParseNode | null = null;
    let i = index;

    if (allowNegative) {
        i = GetLiteralMatch(context, i, "-").NextIndex;
    }

    let i2 = i;
    while (i2 < context.Expression.length && /^\d$/.test(context.Expression[i2])) {
        i2++;
    }

    if (i === i2) {
        return { IntVal: null, ParseNode: parseNode, NextIndex: index };
    }

    i = i2;

    const intVal = context.Expression.substring(index, i);
    parseNode = new ParseNode(ParseNodeType.LiteralInteger, index, i - index);
    return { IntVal: intVal, ParseNode: parseNode, NextIndex: i };
}